mod-source-record-manager-server/src/main/java/org/folio/services/afterprocessing/InstanceProcessingServiceImpl.java

```diff
@@ -75,68 +63,109 @@ public InstanceProcessingServiceImpl(@Autowired JobExecutionSourceChunkDao jobEx
+   * @param params    Okapi connection params
+   * @return future
    */
-  private Future<String> postInstance(Instance instance, OkapiConnectionParams params) {
-    Future<String> future = Future.future();
-    RestUtil.doRequest(params, INVENTORY_URL, HttpMethod.POST, instance).setHandler(responseResult -> {
+  private Future<Void> postInstances(List<Instance> instances, OkapiConnectionParams params) {
+    if (CollectionUtils.isEmpty(instances)) {
+      return Future.succeededFuture();
+    }
+    Future<Void> future = Future.future();
+    RestUtil.doRequest(params, INVENTORY_URL, HttpMethod.POST, instances).setHandler(responseResult -> {
       try {
         if (!RestUtil.validateAsyncResult(responseResult, future)) {
-          LOGGER.error("Error creating new Instance record", future.cause());
+          LOGGER.error("Error creating a new collection of Instances", future.cause());
         } else {
-          if (StringUtils.isNotEmpty(instance.getId())) {
-            String instanceId = instance.getId();
-            future.complete(instanceId);
-          } else {
-            String location = responseResult.result().getResponse().getHeader(INSTANCE_LOCATION_RESPONSE_HEADER);
-            String instanceId = location.substring(location.lastIndexOf('/') + 1);
-            future.complete(instanceId);
-          }
+          future.complete();
         }
       } catch (Exception e) {
-        LOGGER.error("Error during post for new Instance", e);
+        LOGGER.error("Error during post for new collection of Instances", e);
         future.fail(e);
       }
     });
     return future;
   }
 
+  /**
+   * Adds additional custom fields to parsed records and updates parsed records in mod-source-record-storage
+   *
+   * @param recordToInstanceList association between Records and corresponding Instances
+   * @param records              original list of records
+   * @param params               okapi connection params
+   */
+  private void addAdditionalFields(List<Pair<Record, Instance>> recordToInstanceList,
+                                   List<Record> records,
+                                   OkapiConnectionParams params) {
+    if (!CollectionUtils.isEmpty(recordToInstanceList) || !CollectionUtils.isEmpty(records)) {
+      Record.RecordType recordType = getRecordsType(records);
+      if (Record.RecordType.MARC == recordType) {
+        recordToInstanceList.parallelStream().forEach(entry ->
+          additionalInstanceFieldsUtil.addInstanceIdToMarcRecord(entry.getKey(), entry.getValue().getId())
+        );
+        List<Record> recordsFromInstances = recordToInstanceList.parallelStream().map(Pair::getKey).collect(Collectors.toList());
+        updateParsedRecords(recordsFromInstances, params).setHandler(updatedAr -> {
+          if (updatedAr.failed()) {
+            LOGGER.error("Couldn't update parsed records", updatedAr.cause());
+          }
+        });
+      }
+    }
+  }
```

Comment:
I suggest the following version of the method:
the List<Record> records is unnecessary as for me  
```
/**
   * Adds additional custom fields to parsed records and updates parsed records in mod-source-record-storage
   *
   * @param recordToInstanceList association between Records and corresponding Instances
   * @param params               okapi connection params
   */
  private void addAdditionalFields(List<Pair<Record, Instance>> recordToInstanceList, OkapiConnectionParams params) {
    if (CollectionUtils.isEmpty(recordToInstanceList)) {
      return;
    }
    
    if (Record.RecordType.MARC == recordToInstanceList.get(0).getKey().getRecordType()) {
      List<Record> records = recordToInstanceList.parallelStream()
        .peek(it -> additionalInstanceFieldsUtil.addInstanceIdToMarcRecord(it.getKey(), it.getValue().getId()))
        .map(Pair::getKey)
        .collect(Collectors.toList());

      updateParsedRecords(records, params).setHandler(updatedAr -> {
        if (updatedAr.failed()) {
          LOGGER.error("Couldn't update parsed records", updatedAr.cause());
        }
      });
    }
  }
```